lib/spack/docs/packaging_guide.rst

```diff
@@ -6116,3 +6297,82 @@ might write:
+class structure for a package looked like:
+
+.. image:: images/original_package_architecture.png
+   :scale: 60 %
+   :align: center
+
+In this architecture the base class ``AutotoolsPackage`` was responsible for both the metadata
+related to the ``autotools`` build-system (e.g. dependencies or variants common to all packages
+using it), and for encoding the default installation procedure.
+
+In reality, a non-negligible number of packages are either changing their build-system during the evolution of the
+project, or using different build-systems for different platforms. An architecture based on a single class
+allows to deal with these cases only by employing workarounds and possibly by repeating code.
+
+To support a model more adherent to reality, Spack v0.19 changed its internal design by extracting
+the attributes and methods related to building a software into a separate hierarchy:
+
+.. image:: images/builder_package_architecture.png
+   :scale: 60 %
+   :align: center
+
+In this new format each ``package.py`` contains one ``*Package`` class that gathers all the metadata,
+and one or more ``*Builder`` classes that encode the installation procedure. A specific builder object
+is created just before the software is built, so at a time where Spack knows which build-system needs
+to be used for the current installation, and receives a ``package`` object during initialization.
+
+^^^^^^^^^^^^^^^^^^^^^^^^
+``build_system`` variant
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+To allow imposing conditions based on the build-system, each package must a have ``build_system`` variant,
+which is usually inherited from base classes. This variant allows both to write metadata that is conditional
+to the build-system:
+
+.. code-block:: python
+
+   with when("build_system=cmake"):
+       depends_on("cmake", type="build")
+
+and also to select a specific build system from a spec literal, e.g. like in the following command:
+
+.. code-block:: console
+
+   $ spack install arpack-ng build_system=autotools
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Compatibility with single-class format
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Internally, Spack always uses builders to perform operations related to the installation of a specific software.
+The builders are created in the ``spack.builder.create`` function
+
+.. literalinclude:: _spack_root/lib/spack/spack/builder.py
+   :pyobject: create
+
+To achieve backward compatibility with the single-class format Spack creates in this function a special
+"adapter builder", if no custom builder is detected in the recipe:
+
+.. image:: images/adapter.png
+   :scale: 60 %
+   :align: center
+
+Overall the role of the adapter is to route access to attributes of methods first through the ``*Package``
+hierarchy, and then back to the base class builder. This is schematically shown in the diagram above, where
+the adapter role is to "emulate" a method resolution order like the one represented by the red arrows.
```

Comment:
A lot of the stuff added here (history, design, implementation) is great, but feels like it should be somewhere else. The Packaging Guide is becoming quite bloated. But we can refactor this in a separate PR. I would rather document this somewhere than nowhere.

Reply:
Yes, working on docs in a separate PR would be simpler. I have a few small PRs that should follow this one.